stream-tests/src/test/scala/org/apache/pekko/stream/scaladsl/FlowSplitAfterSpec.scala

```diff
@@ -250,10 +248,10 @@ class FlowSplitAfterSpec extends StreamSpec("""
       upstreamSubscription.sendNext(6)
       substreamPuppet1.expectNext(6)
       substreamPuppet1.expectComplete()
+      upstreamSubscription.sendNext(7)
```

Comment:
It seems like this is a genuine mistake when the test was originally written (remember that this test never passed, it was written for the future when supervision strategy propagation would be implemented which occurred due to https://github.com/apache/incubator-pekko/pull/252).

Logically when you read the test, its expected that when you do `expectNext` that a previously an element should have been sent with `sendNext` and its also the same in  the equivalent [`groupBy` test](https://github.com/apache/incubator-pekko/blob/c44c0b7cbdab11d85176cfe062288fdcba16c56a/stream-tests/src/test/scala/org/apache/pekko/stream/scaladsl/FlowGroupBySpec.scala#L306-L308)

Reply:
I thought about this again, I may be wrong and the test is indeed valid as its written right now because its `splitAfter` so its its meant to be skipping the next element due to exception being thrown, in which case we have to drop one element and then do `pull(in)`?

Reply:
Yes, the element which case the exception is dropped now.

Reply:
> Yes, the element which case the exception is dropped now.

So to confirm, the current PRs implementation is correct and the test as originally written was wrong?

Reply:
As the `downstreamSubscription.request(100)` , I think the orign test should be right too, but its a little strange now ,that we need send the `next(7)` to make the next sub source be ready.

I think we should call 
```scala
      private def pushSubstreamSource(): Unit = {
        push(out, Source.fromGraph(substreamSource.source))
        scheduleOnce(SubscriptionTimer, timeout)
        substreamWaitingToBePushed = false
      }
```

now inside the `SubstreamHandler#onPush` when the decision is `SplitAfter` too to keep the behavior the same, as it just a dummy Source.

Reply:
So if I change the logic to

```scala
          case NonFatal(ex) =>
            decider(ex) match {
              case Supervision.Resume  =>
                if (decision == SplitAfter)
                  pushSubstreamSource()
                else
                  pull(in)
              case Supervision.Stop    => onUpstreamFailure(ex)
              case Supervision.Restart => onUpstreamFailure(ex) // TODO implement restart?
            }
```

This test now fails earlier on in the happy path, specifically at https://github.com/apache/incubator-pekko/blob/df3b9fccde6b7fbe7829703996379a055bc27b42/stream-tests/src/test/scala/org/apache/pekko/stream/scaladsl/FlowSplitAfterSpec.scala#L243

Maybe the test really was just originally written incorrectly?

Reply:
I was reviewing this test. Isn't the `val substream2 = subscriber.expectNext()` expecting the new stream element (substream2) triggered by the `sendNext(6)` instruction? why do we need to move the `sendNext(7)` above?

Reply:
> I was reviewing this test. Isn't the val substream2 = subscriber.expectNext() expecting the new stream element (substream2) triggered by the sendNext(6) instruction? why do we need to move the sendNext(7) above?

Well the test doesn't pass otherwise and if you see the equivalent tests both in [`FlowSplitWhenSpec`](https://github.com/apache/incubator-pekko/blob/df3b9fccde6b7fbe7829703996379a055bc27b42/stream-tests/src/test/scala/org/apache/pekko/stream/scaladsl/FlowSplitWhenSpec.scala#L246-L278) and [`FlowGroupBySpec`](https://github.com/apache/incubator-pekko/blob/c44c0b7cbdab11d85176cfe062288fdcba16c56a/stream-tests/src/test/scala/org/apache/pekko/stream/scaladsl/FlowGroupBySpec.scala#L281-L323), its written in the same way that this PR changes it to.

There also doesn't appear to be an obvious solution to make the test as it was originally written to pass, hence why considering everything it could just be that the test was mistakingly written incorrectly (which is quite probably considering the test was written just as an example, it was never executed/passing)

Reply:
@mdedetrich in the test, it requests 10, so the SubSource is expected to be there after 6.

Reply:
> @mdedetrich in the test, it requests 10, so the SubSource is expected to be there after 6.

But its also sending a completion because of this condition `if (elem == 3) throw exc else elem % 3 == 0` (i.e. `6 % 3` == 0 so the split happens) which is why the test is failing with

```
Expected OnNext(_), yet no element signaled during 3 seconds
java.lang.AssertionError: Expected OnNext(_), yet no element signaled during 3 seconds
	at org.apache.pekko.stream.testkit.TestSubscriber$ManualProbe.expectNext(StreamTestKit.scala:398)
	at org.apache.pekko.stream.testkit.TestSubscriber$ManualProbe.expectNext(StreamTestKit.scala:389)
	at org.apache.pekko.stream.scaladsl.FlowSplitAfterSpec.$anonfun$new$11(FlowSplitAfterSpec.scala:251)
```

i.e. at `val substream2 = subscriber.expectNext()` because there isn't going to be a `expectNext` due to the completion?

Reply:
**1. `val substream2 = subscriber.expectNext()`, the `subscriber` is not expecting a `complete`,  otherwise, the call `sendNext(7)` will cause issue because the origin source is already completed.

The problem here is when the `stream2` be generated, I expected the behavior the same, just after the `6 % 3 == 0`, wdyt @samueleresca @mdedetrich**

Reply:
> **1. `val substream2 = subscriber.expectNext()`, the `subscriber` is not expecting a `complete`, otherwise, the call `sendNext(7)` will cause issue because the origin source is already completed.
> 
> The problem here is when the `stream2` be generated, I expected the behavior the same, just after the `6 % 3 == 0`, wdyt @samueleresca @mdedetrich**

I am not saying that this is wrong, just that its a completely separate issue from what this PR is changing/solving. Or to put it differently, this behaviour is the same as current Pekko since at this point i.e.

```scala
upstreamSubscription.sendNext(6)
substreamPuppet1.expectNext(6)
substreamPuppet1.expectComplete()
upstreamSubscription.sendNext(7)
val substream2 = subscriber.expectNext()
val substreamPuppet2 = StreamPuppet(substream2.runWith(Sink.asPublisher(false)))
substreamPuppet2.request(10)
substreamPuppet2.expectNext(7)

upstreamSubscription.sendComplete()
subscriber.expectComplete()
substreamPuppet2.expectComplete()
```

we are past recovering from an exception. The critical part of the test specifically dealing with recovering from exception and resuming is

```scala
upstreamSubscription.sendNext(3)
upstreamSubscription.sendNext(4)
substreamPuppet1.expectNext(4) // note that 3 was dropped
```

And this part is completely unchanged from how the test was originally written.

Given that, I think it makes sense to file a separate issue if this behaviour about completing after a split needs to change and hence to tackle it separately

Reply:
Thanks folks, I appreciate the detailed explanation. And yes, I was diverting away (not on purpose) from the original goal of the test 

> Note that at this point, we are dealing with the normal logic of SplitWhen. The new functionality of skipping 
> exceptions with SupervisionDecider.resume occurs earlier when elem == 3 so we are past that and just testing happy > path logic of SplitAfter which is not changed at all (and it shouldn't be either!).

looks good to me